File: problems/3483-unique-3-digit-even-numbers/solution.erl
-spec total_numbers(Digits :: [integer()]) -> integer().
total_numbers(Digits) ->
    FreqMap = lists:foldl(fun(Digit, Acc) -> Acc#{Digit => maps:get(Digit, Acc, 0) + 1} end, #{}, Digits),
    count_by_hundreds(maps:keys(FreqMap), FreqMap, 0).

count_by_hundreds([], _FreqMap, Count) ->
    Count;
count_by_hundreds([0 | R], FreqMap, Count) ->
    count_by_hundreds(R, FreqMap, Count);
count_by_hundreds([Hundreds | R], FreqMap, Count) ->
    Remaining = FreqMap#{Hundreds => maps:get(Hundreds, FreqMap) - 1},
    count_by_hundreds(R, FreqMap, Count + count_by_units(maps:keys(FreqMap), Remaining, 0)).

count_by_units([], _FreqMap, Acc) ->
    Acc;
count_by_units([Unit|R], FreqMap, Acc) ->
    case (Unit rem 2 == 1) orelse maps:get(Unit, FreqMap) of
        true ->
            count_by_units(R, FreqMap, Acc);
        0 ->
            count_by_units(R, FreqMap, Acc);
        Freq ->
            Count = count_available_middles(maps:to_list(FreqMap#{Unit => Freq - 1})),
            count_by_units(R, FreqMap, Acc + Count)
    end.

count_available_middles(L) ->
    length(lists:filter(fun({_, V}) -> V >= 1 end , L)).

File: problems/3483-unique-3-digit-even-numbers/solution.ex
defmodule Solution do
  @spec total_numbers(digits :: [integer]) :: integer
  def total_numbers(digits) do
    freq_map =
      Enum.reduce(digits, %{}, fn digit, acc ->
        Map.update(acc, digit, 1, &(&1 + 1))
      end)

    count_by_hundreds(Map.keys(freq_map), freq_map, 0)
  end

  defp count_by_hundreds([], _freq_map, count), do: count

  defp count_by_hundreds([0 | rest], freq_map, count) do
    count_by_hundreds(rest, freq_map, count)
  end

  defp count_by_hundreds([hundreds | rest], freq_map, count) do
    remaining = Map.update(freq_map, hundreds, 0, &(&1 - 1))
    count_by_hundreds(rest, freq_map, count + count_by_units(remaining))
  end

  defp count_by_units(freq_map) do
    freq_map
    |> Map.keys()
    |> Enum.reduce(0, fn units, count ->
      cond do
        Map.get(freq_map, units) == 0 ->
          count

        rem(units, 2) != 0 ->
          count

        true ->
          remaining = Map.update(freq_map, units, 0, &(&1 - 1))
          count + count_available_middles(remaining)
      end
    end)
  end

  defp count_available_middles(freq_map) do
    freq_map
    |> Map.values()
    |> Enum.count(&(&1 > 0))
  end
end
